### DISCUSSION TITLE
UART odd and even parity

### DISCUSSION THREAD

**Discussion by msawicki-poland** [Pi4J/pi4j#272] [General]
The Raspberry Pi 4 board with the Raspberry Pi OS and Java 17 is connected via UART to the computer (macOS). Communication parameters are 9600, 8 data bits, 1 stop bit. On Raspberry and macOS, minicom is running and communication works in both dir for no parity and for even and odd parity. In the second step, a program written on the Raspberry was launched:

```
public class Fan
{
    public static void main(String[] args) throws Exception
    {
	var pi4j=Pi4J.newAutoContext();
	var serial=pi4j.create(Serial.newConfigBuilder(pi4j)
		.use_9600_N81()
		.dataBits_8()
		.parity(Parity.NONE) //or ODD or EVEN
		.stopBits(StopBits._1)
		.flowControl(FlowControl.NONE)
		.id("serial")
		.device("/dev/ttyAMA1")
		.provider("pigpio-serial")
		.build());

	serial.open();

	byte[] command = "at\r".getBytes();
	serial.write(command);
    }
}
```
for no parity it works, i.e. macOS receives characters correctly "at\r". On the other hand, for even and odd parities, the '\r' character (0x0D) is changed to 0x0C or 0xC3 but "at" is sent correctly.

**Comment by taartspi** [comment]
I am updating this post as after more investigation the following details are all correct, however the underlying PIGPIO Library does not support the required parity.  For the parity options to function it would require PIGPIO updates and pi4j -v2 updates to use that functionality.

serOpen This function opens a serial device at a specified baud rate and with specified flags. The device name must start with /dev/tty or /dev/serial. 
   the serial device to open
  baud: the baud rate in bits per second, see below
  serFlags: 0
Returns a handle (>=0) if OK, otherwise PI_NO_HANDLE, or PI_SER_OPEN_FAILED.
The baud rate must be one of 50, 75, 110, 134, 150, 200, 300, 600, 1200, 1800, 2400, 4800, 9600, 19200, 38400, 57600, 115200, or 230400.

**No flags are currently defined. This parameter should be set to zero.**




If you are using pin 8 and 10, GPIO14 and GPIO15 read on.  The BCM2711 device has six UARTs. One mini UART (UART1) and five PL011 UARTs (UART0, UART2, UART3, UART4 &UART5).

  Those pins are the mini-com uart--   Does not have the parity you wanted.
**The mini UART has the following features:**
•
7-bit or 8-bit operation
•
1 start and 1 stop bit
• No parities
•
Break generation
•
8 symbols deep FIFOs for receive and transmit
•
SW controlled RTS, SW readable CTS
•
Auto flow control with programmable FIFO level
•
16550 like registers
•
Baudrate derived from system clock
**This is a mini UART and it does NOT have the following** capabilities:
•
Break detection
•
Framing errors detection
•
Parity bit
•
Receive Time-out interrupt
•
DCD, DSR, DTR or RI signals




**You need to use one of the ALT uarts PL011**
The UARTs provide:
•
Separate 32x8 transmit and 32x12 receive FIFO memory.
•
Programmable baud rate generator.
•
Standard asynchronous communication bits (start, stop and parity). These are added prior to transmission and
removed on reception.
•
False start bit detection.
•
Line break generation and detection.
•
Support of the modem control functions CTS and RTS. However DCD, DSR, DTR, and RI are not supported.
•
Programmable hardware flow control.
•
Fully-programmable serial interface characteristics:
◦ data can be 5, 6, 7, or 8 bits.
◦ even, odd, stick, or no-parity bit generation and detection.
◦ 1 or 2 stop bit generation.
◦ baud rate generation, up to UARTCLK/16.


If you are already using an ALT uart let me know and I will investigate more

https://datasheets.raspberrypi.com/bcm2711/bcm2711-peripherals.pdf

**Comment by taartspi** [comment]
See above comment for full details.
The underlying PIGPIO Library does not support the required parity.  For the parity options to function it would require PIGPIO updates and pi4j -v2 updates to use that functionality.

At present time you would be required to write the code to 'bit bang' GPIO pins yourself to achieve this parity.
